### PR TITLE
KotlinGenerator supports custom options for message, field, enum

### LINKED ISSUE #1271: Support custom options in Kotlin
See [custom_options.proto](https://github.com/square/wire/blob/master/wire-tests/src/commonTest/proto/java/custom_options.proto).

### PR DESCRIPTION
**Pull request by reneeleung**
closes #1271

### CODE REVIEW

File: wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt [hunk 1]
```diff
@@ -59,26 +59,18 @@ import com.squareup.wire.ProtoWriter
 import com.squareup.wire.WireEnum
 import com.squareup.wire.WireField
 import com.squareup.wire.WireRpc
-import com.squareup.wire.schema.EnclosingType
-import com.squareup.wire.schema.EnumType
-import com.squareup.wire.schema.Field
-import com.squareup.wire.schema.MessageType
-import com.squareup.wire.schema.OneOf
+import com.squareup.wire.schema.*
```
Comment (oldergod) [inline wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt]: let's be explicit with our imports

File: wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt [hunk 2]
```diff
@@ -59,26 +59,18 @@ import com.squareup.wire.ProtoWriter
...
+import com.squareup.wire.schema.Options.Companion.FIELD_OPTIONS
+import com.squareup.wire.schema.Options.Companion.MESSAGE_OPTIONS
 import kotlinx.coroutines.Deferred
 import kotlinx.coroutines.channels.ReceiveChannel
 import kotlinx.coroutines.channels.SendChannel
 import okio.ByteString
 import okio.ByteString.Companion.encode
 import java.math.BigInteger
-import java.util.Locale
+import java.util.*
```
Comment (oldergod) [inline wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt]: explicit

**Comment by oldergod** [comment]
Please, rebase master and `./gradlew generateTests` before pushing again.

File: wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt [hunk 3]
```diff
@@ -81,12 +73,12 @@ import java.math.BigInteger
...
-  private val emitAndroid: Boolean,
-  private val javaInterOp: Boolean,
-  private val rpcCallStyle: RpcCallStyle,
-  private val rpcRole: RpcRole
+        val schema: Schema,
+        private val nameToKotlinName: Map<ProtoType, ClassName>,
+        private val emitAndroid: Boolean,
+        private val javaInterOp: Boolean,
+        private val rpcCallStyle: RpcCallStyle,
+        private val rpcRole: RpcRole
```
Comment (oldergod) [inline wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt]: indent is off

File: wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt [hunk 4]
```diff
@@ -1168,10 +1271,10 @@ class KotlinGenerator private constructor(
           addCode("\n⇤}\n") // close the block
         }
         .build()
-    return TypeSpec.companionObjectBuilder()
-        .addFunction(fromValue)
-        .addProperty(generateEnumAdapter(message))
-        .build()
+
+    return companionObjectBuilder.addFunction(fromValue)
+            .addProperty(generateEnumAdapter(message))
```
Comment (Egorand) [inline wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt]: nit: revert indentation changes, 4 spaces is correct.

File: wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt [hunk 5]
```diff
@@ -1148,10 +1245,16 @@ class KotlinGenerator private constructor(
           .build())
     }
 
-    return builder.build()
+    return builder.primaryConstructor(primaryConstructor.build())
+            .build()
```
Comment (Egorand) [inline wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt]: nit: 4 space indentation.

File: wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt [hunk 6]
```diff
@@ -1117,25 +1190,49 @@ class KotlinGenerator private constructor(
...
+    val allOptionFieldsBuilder = mutableSetOf<ProtoMember>()
+
+    // Each enum constant option requires a constructor parameter and property
+    message.constants.forEach { constant ->
+      constant.options.map.keys.forEach { protoMember ->
+        if (allOptionFieldsBuilder.add(protoMember)) {
+          val optionField = schema.getField(protoMember)
+          primaryConstructor.addParameter(optionField.name, optionField.typeName)
+          builder.addProperty(PropertySpec.builder(optionField.name, optionField.typeName)
+                  .initializer(optionField.name)
```
Comment (Egorand) [inline wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt]: nit: 4 space indentation.

File: wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt [hunk 7]
```diff
@@ -1117,25 +1190,49 @@ class KotlinGenerator private constructor(
 
     val valueName = nameAllocator["value"]
 
+    val primaryConstructor = FunSpec.constructorBuilder()
+            .addParameter(valueName, Int::class, OVERRIDE)
```
Comment (Egorand) [inline wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt]: nit: 4 space indentation.

File: wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt [hunk 8]
```diff
@@ -1047,6 +1099,27 @@ class KotlinGenerator private constructor(
...
+          if (!empty) add(",")
+          add("\n⇥%L = %L⇤", nameAllocator[optionField], defaultFieldInitializer(optionField.type!!, entry.value!!))
+          empty = false
+        }
+      }
+      add("\n)")
+      if (empty) return null
+    }
+    return PropertySpec.builder(fieldName, optionsType.typeName)
+            .initializer(initializer)
```
Comment (Egorand) [inline wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt]: nit: 4 space indentation.

File: wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt [hunk 9]
```diff
@@ -1294,12 +1397,17 @@ class KotlinGenerator private constructor(
         ProtoType.SINT64 to LONG,
         ProtoType.STRING to String::class.asClassName(),
         ProtoType.UINT32 to INT,
-        ProtoType.UINT64 to LONG
+        ProtoType.UINT64 to LONG,
+        FIELD_OPTIONS to ClassName("com.google.protobuf", "FieldOptions"),
```
Comment (Egorand) [inline wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt]: Nice!

File: wire-tests/src/commonTest/proto-kotlin/com/squareup/wire/protos/custom_options/FooBar.kt [hunk 10]
```diff
@@ -0,0 +1,478 @@
...
+    fred: List<Float> = this.fred,
+    daisy: Double? = this.daisy,
+    nested: List<FooBar> = this.nested,
+    ext: FooBarBazEnum? = this.ext,
+    rep: List<FooBarBazEnum> = this.rep,
+    unknownFields: ByteString = this.unknownFields
+  ): FooBar = FooBar(foo, bar, baz, qux, fred, daisy, nested, ext, rep, unknownFields)
+
+  companion object {
+    val FIELD_OPTIONS_FOO: FieldOptions = FieldOptions(
```
Comment (Egorand) [inline wire-tests/src/commonTest/proto-kotlin/com/squareup/wire/protos/custom_options/FooBar.kt]: Looks good! We should fixup indentation a bit, shouldn't be that hard - can help.

File: wire-tests/src/commonTest/proto-kotlin/com/squareup/wire/protos/kotlin/DeprecatedEnum.kt [hunk 11]
```diff
@@ -5,23 +5,25 @@ package com.squareup.wire.protos.kotlin
...
+import kotlin.Boolean
 import kotlin.Deprecated
 import kotlin.Int
 import kotlin.jvm.JvmField
 import kotlin.jvm.JvmStatic
 
 enum class DeprecatedEnum(
-  override val value: Int
+  override val value: Int,
+  val deprecated: Boolean?
```
Comment (Egorand) [inline wire-tests/src/commonTest/proto-kotlin/com/squareup/wire/protos/kotlin/DeprecatedEnum.kt]: @oldergod was this what you were seeing for Java as well? Feels like a bug. @swankjesse do you know why we're printing out `deprecated` as an enum field in addition to `@Deprecated`?

File: wire-tests/src/jvmKotlinInteropTest/proto-kotlin/com/squareup/wire/protos/kotlin/simple/SimpleMessage.kt [hunk 12]
```diff
@@ -599,18 +600,23 @@ class SimpleMessage(
...
 
-    BAZ(3),
+    BAZ(3, null),
 
     @Deprecated(message = "BUZ is deprecated")
-    BUZ(3);
+    BUZ(3, true);
 
     companion object {
+      val ENUM_OPTIONS: EnumOptions = EnumOptions(
```
Comment (Egorand) [inline wire-tests/src/jvmKotlinInteropTest/proto-kotlin/com/squareup/wire/protos/kotlin/simple/SimpleMessage.kt]: In interop mode we should generate `@JvmField`s for those, let's do this in a followup.

File: wire-tests/src/commonTest/proto-kotlin/com/squareup/wire/protos/kotlin/DeprecatedEnum.kt (hunk 11 above)
Comment (oldergod) [reply wire-tests/src/commonTest/proto-kotlin/com/squareup/wire/protos/kotlin/DeprecatedEnum.kt]: Java does that indeed. https://github.com/square/wire/pull/1315/files#diff-e253bdd1b1acf18fd5cdeb67a950ecc8R526
Renee and I talked and we decided to first do as what Java is already doing. If that's a bug, let's fix it later in one go.

Comment (Egorand) [reply wire-tests/src/commonTest/proto-kotlin/com/squareup/wire/protos/kotlin/DeprecatedEnum.kt]: SGTM!

File: wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt [hunk 13]
```diff
@@ -257,6 +261,7 @@ class KotlinGenerator private constructor(
             newName("reader", "reader")
             newName("Builder", "Builder")
             newName("builder", "builder")
+            newName("MESSAGE_OPTIONS", "MESSAGE_OPTIONS")
```
[the remark this answers is not in the document]
Comment (reneeleung) [reply wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt]: we have name allocator for each field already, but will add one for `ENUM_OPTIONS` in follow-up

File: wire-tests/src/commonTest/proto-kotlin/com/squareup/wire/protos/kotlin/DeprecatedEnum.kt (hunk 11 above)
Comment (reneeleung) [reply wire-tests/src/commonTest/proto-kotlin/com/squareup/wire/protos/kotlin/DeprecatedEnum.kt]: #1312

File: wire-tests/src/jvmKotlinInteropTest/proto-kotlin/com/squareup/wire/protos/kotlin/simple/SimpleMessage.kt (hunk 12 above)
Comment (reneeleung) [reply wire-tests/src/jvmKotlinInteropTest/proto-kotlin/com/squareup/wire/protos/kotlin/simple/SimpleMessage.kt]: sounds good!

**Pull request merged**